### ISSUE TITLE
Distributed.matvec rebuilds and re-transfers the full embedding tensor on every call, defeating sharding

### ISSUE THREAD

**Issue by sachncs** [sachncs/laker#18]
## Problem

`Distributed.matvec` defeats the purpose of multi-GPU sharding by materialising the full concatenated embedding tensor on every matvec call and then re-transferring it to each device on every iteration of the inner loop. The cross-device traffic is O(devices × n) per call, where the entire point of sharding was to make each device only see its O(n / devices) shard.

## Evidence

`laker/distributed.py:90-123`:

```python
def matvec(self, x: torch.Tensor) -> torch.Tensor:
    if self.single:
        return self.local_op.matvec(x)

    x_m = x.to(self.master)
    full = torch.cat(                                      # rebuilt every call
        [op.embeddings.to(self.master) for op in self.ops],
        dim=0,
    )
    outs = []
    start = 0
    for op in self.ops:
        device = op.embeddings.device
        local = op.embeddings
        sz = local.shape[0]
        end = start + sz

        d_t = x_m.to(device)                               # per-device x transfer
        full_d = full.to(device)                           # per-device full tensor transfer

        chunk_size = 8192
        local_out = self.lam * d_t
        for j in range(0, self.size, chunk_size):
            j_end = min(j + chunk_size, self.size)
            gb = local @ full_d[j:j_end].T                 # uses full_d, not local shard
            exp_safe(gb, out=gb, skip=False)
            ...
        outs.append(local_out[start:end].to(self.master))
        start = end

    return torch.cat(outs, dim=0)
```

Three problems:

1. `torch.cat([op.embeddings.to(self.master) for op in self.ops], dim=0)` allocates and concatenates `n` floats every call (line 95-98).
2. `full_d = full.to(device)` then transfers the entire concatenated tensor to each device (line 108) — every device receives the full `n`-row matrix, not just the slice it should see.
3. The chunked matmul `local @ full_d[j:j_end].T` reads the *full* `j:j_end` slice of the global `full_d`, which means the local shard `local` (rows `start:end`) reads the *entire* `n × d` matrix per device.

The actual work performed across all devices is `O(devices² × n² / devices²) = O(n²)` of matmul, i.e. no savings from sharding. The expected pattern is to compute the *cross-block* terms: each device holds rows `start:end` and computes `(local @ remote.T)` for each remote shard, summed across devices. The current code skips that structure and re-implements the global matvec on each device.

## Root Cause

`Distributed.matvec` was written as a thin wrapper over the local `Exact` operator and treats the device list as a list of "extra copies" rather than a partition. The cleanest fix is to keep a single concatenated embedding tensor (cached at construction time) and a precomputed list of `(start, end, device)` slices; then each device does `local @ remote_shard.T` and a `sum` over remotes.

## Impact

For `n=10_000`, `devices=2`, and a single matvec: instead of one device moving O(n/2) data and doing O(n²/4) flops, the code moves O(2n) data and does O(n²) flops. Net effect: the multi-GPU path is *slower* than the single-GPU path for any non-trivial `n`, and the README's "Multi-GPU distributed matvec" claim is misleading.

`tests/test_distributed.py` does not exercise the multi-GPU path (CI only has CPU/macOS runners), so this is not caught in CI.

## Recommendation

Refactor `Distributed.matvec` so that:

1. `self.full` is computed once in `__init__` and stored on the master device.
2. Each device's slice `[start:end]` is cached as `self.shards` (already is via `op.embeddings`).
3. For each pair `(local_device, remote_device)`, schedule an async matmul of `local @ remote.T`; the master sums the partial results.
4. Remove the `full_d = full.to(device)` call inside the inner loop — that is the worst offender.

Add a test that measures wall-clock time for `Distributed.matvec` on a synthetic 2-device setup and asserts it is within a constant factor of the single-device path, plus a small unit test that asserts the output is correct.

## Acceptance Criteria

- [ ] `Distributed.matvec` performs at most O(n × n / devices) flops in the steady state.
- [ ] The concatenated embedding tensor is built once, not per call.
- [ ] A correctness test asserts `Distributed(A) @ v == Exact(A) @ v` for a multi-shard configuration.
- [ ] A regression benchmark under `benchmarks/` exercises the multi-GPU path on 2 simulated devices.

## Verification

Run the new correctness test and the new benchmark; the benchmark should report a clear speedup over the current implementation on `n ≥ 5000`.

## Dependencies

None.

## Risk

Medium. `Distributed` is exercised through `Laker(distributed=True)` which is the recommended path for large problems. Behaviour must remain numerically equivalent (`assert torch.allclose(out, expected, atol=1e-4)`).

**Comment by sachncs** [comment]
Resolved in PR #32.
